lib/Core/Executor.cpp

```diff
@@ -3888,17 +4012,73 @@ void Executor::executeCheckPersistence(ExecutionState &state,
   assert(os);
   ObjectState *wos = state.addressSpace.getWriteable(mo, os);
   PersistentState *ps = dyn_cast<PersistentState>(wos);
+  assert(ps);
+
+  #if 1
+  ConstraintManager orig = state.constraints;
+
+  for (const ref<Expr> &sliceConstraint : ps->getConstraints()) {
+    // fprintf(stderr, "check slice for %p\n", (void*)mo->address);
+    // sliceConstraint->dump();
+    ConstraintManager cm = orig;
+    cm.addConstraint(sliceConstraint);
+    state.constraints = cm;
+
+    StatePair isPersisted = fork(state, ps->isPersistedUnconstrained(), true);
+
+    // If there's a state where it's not definitely persisted, terminate it.
+    ExecutionState *notPersisted = isPersisted.second;
```

Comment:
I came back here to make a comment about how I prefer the look of just calling `PersistentState::mustBePersistent` from here and passing any additional information back through out-parameters. However, looking at it again, there seems to be more going on here. In fact, I like the use of `fork()` here instead of just one big boolean function to check the whole thing. A neat side effect is that the `notPersisted` state's constraints contain the particular slice that was found to be not persisted, which could help the solver when we try to evaluate which cache lines are dirty.

So, I say let's do _all_ the checking logic here in the Executor. Let's create the symbolic offset here using `arrayCache->CreateArray()` rather than having it be a member variable of `PersistentState` (if you use the same name and size each time, it won't create a new Array, _a la_ "cache"). Also, I think we could get away with creating the slice constraints on the fly (they're cheap wrappers) rather than storing those as member variables. Then the "slice size" could be a configurable Executor parameter.

This brings us back to the good spot of not having `PersistentState` be responsible for a lot of logic.

Reply:
I'm not sure if I'm convinced. To me, these things are more like properties of the `PersistentState`, so to me it makes more sense to stash them in the `PersistentState`. But, I agree that it would parse out some logic from the `PersistentState`.

I'll put a TODO for potential refactoring.